pkg/tun/tun_darwin.go

```diff
@@ -0,0 +1,247 @@
+		}
+		return err
+	})
+}
+
+// Read reads as many bytes as possible into the given buffer.Data and returns the
+// number of bytes actually read
+func (t *Device) Read(into *buffer.Data) (int, error) {
+	n, err := t.File.Read(into.Raw())
+	if n >= buffer.PrefixLen {
+		n -= buffer.PrefixLen
+	}
+	return n, err
+}
+
+// Write writes bytes from the given buffer.Data and returns the number of bytes
+// actually written.
+func (t *Device) Write(from *buffer.Data) (int, error) {
+	raw := from.Raw()
+	if len(raw) <= buffer.PrefixLen {
+		return 0, unix.EIO
+	}
+
+	ipVer := raw[buffer.PrefixLen] >> 4
+	switch ipVer {
+	case ipv4.Version:
+		raw[3] = unix.AF_INET
+	case ipv6.Version:
+		raw[3] = unix.AF_INET6
+	default:
+		return 0, errors.New("unable to determine IP version from packet")
+	}
+	n, err := t.File.Write(raw)
+	return n - buffer.PrefixLen, err
+}
+
+// Address structure for the SIOCAIFADDR ioctlHandle request
+//
+// See https://www.unix.com/man-page/osx/4/netintro/
+type addrIfReq struct {
+	name [unix.IFNAMSIZ]byte
+	addr unix.RawSockaddrInet4
+	dest unix.RawSockaddrInet4
+	mask unix.RawSockaddrInet4
+}
+
+// Address structure for the SIOCAIFADDR_IN6 ioctlHandle request
+//
+// See https://www.unix.com/man-page/osx/4/netintro/
+type addrIfReq6 struct {
+	name           [unix.IFNAMSIZ]byte
+	addr           unix.RawSockaddrInet6
+	dest           unix.RawSockaddrInet6
+	mask           unix.RawSockaddrInet6
+	flags          int32 // nolint:structcheck
+	expire         int64 // nolint:structcheck
+	preferred      int64 // nolint:structcheck
+	validLifeTime  uint32
+	prefixLifeTime uint32
+}
+
+// SIOCAIFADDR_IN6 is the same ioctlHandle identifier as unix.SIOCAIFADDR adjusted with size of addrIfReq6
+const SIOCAIFADDR_IN6 = (unix.SIOCAIFADDR & 0xe000ffff) | (uint(unsafe.Sizeof(addrIfReq6{})) << 16)
+const ND6_INFINITE_LIFETIME = 0xffffffff
+
+// SIOCDIFADDR_IN6 is the same ioctlHandle identifier as unix.SIOCDIFADDR adjusted with size of addrIfReq6
+const SIOCDIFADDR_IN6 = (unix.SIOCDIFADDR & 0xe000ffff) | (uint(unsafe.Sizeof(addrIfReq6{})) << 16)
+
+func addrToIp4(subnet *net.IPNet, to net.IP) (*net.IPNet, net.IP, bool) {
+	if to4 := to.To4(); to4 != nil {
+		if dest4 := subnet.IP.To4(); dest4 != nil {
+			if _, bits := subnet.Mask.Size(); bits == 32 {
+				return &net.IPNet{IP: dest4, Mask: subnet.Mask}, to4, true
+			}
+		}
+	}
+	return nil, nil, false
+}
+
+func (t *Device) setAddr(subnet *net.IPNet, to net.IP) error {
+	if sub4, to4, ok := addrToIp4(subnet, to); ok {
+		return withSocket(unix.AF_INET, func(fd int) error {
+			ifreq := &addrIfReq{
+				addr: unix.RawSockaddrInet4{Len: 16, Family: unix.AF_INET},
+				dest: unix.RawSockaddrInet4{Len: 16, Family: unix.AF_INET},
+				mask: unix.RawSockaddrInet4{Len: 16, Family: unix.AF_INET},
+			}
+			copy(ifreq.name[:], t.name)
+			copy(ifreq.addr.Addr[:], sub4.IP)
+			copy(ifreq.mask.Addr[:], sub4.Mask)
+			copy(ifreq.dest.Addr[:], to4)
```

Comment:
IMO, it would be a big help to readability to define some `Cast{IfName,IPv{4,6}{,Mask}` helper functions to just define these as struct literals.

Reply:
It's not clear to me what you mean here. Can you please give an example?